backend/typescript/rest/volunteerRoutes.ts

```diff
@@ -1,70 +1,128 @@
+  - userId, through query param (ex. /volunteer/?userId=1)
+*/
+volunteerRouter.get("/:volunteerId?", async (req, res) => {
+  const { volunteerId } = req.params;
+  const { userId } = req.query;
 
-volunteerRouter.get("/:volunteerID", async (req, res) => {
-  const { volunteerID } = req.params;
   const contentType = req.headers["content-type"];
 
-  if (!volunteerID) {
+  if (volunteerId && userId) {
     await sendResponseByMimeType(res, 400, contentType, [
       {
-        error: "Cannot query by missing volunteerID.",
+        error: "Cannot query by both volunteerId and userId.",
       },
     ]);
     return;
   }
 
-  if (volunteerID) {
-    if (typeof volunteerID !== "string") {
+  if (!volunteerId && !userId) {
+    try {
+      const volunteers = await volunteerService.getVolunteers();
+      await sendResponseByMimeType(res, 200, contentType, volunteers);
+    } catch (error) {
+      await sendResponseByMimeType(res, 500, contentType, [
+        {
+          error: getErrorMessage(error),
+        },
+      ]);
+    }
+    return;
+  }
+
+  if (volunteerId) {
+    if (typeof volunteerId !== "string") {
       res
         .status(400)
-        .json({ error: "volunteerID query parameter must be a string." });
+        .json({ error: "volunteerId query parameter must be a string." });
     } else {
       try {
-        const volunteer = await volunteerService.getVolunteerByID(volunteerID);
+        const volunteer = await volunteerService.getVolunteerById(volunteerId);
         res.status(200).json(volunteer);
       } catch (error: unknown) {
         res.status(500).json({ error: getErrorMessage(error) });
       }
     }
   }
-});
 
-volunteerRouter.put("/:volunteerID", async (req, res) => {
-  const { id } = req.query;
+  if (userId) {
+    if (typeof userId !== "string") {
+      res
+        .status(400)
+        .json({ error: "userId query parameter must be a string" });
+      return;
+    }
+    try {
+      const volunteer = await volunteerService.getVolunteerByUserId(userId);
+      res.status(200).json(volunteer);
+    } catch (error) {
+      res.status(500).json({ error: getErrorMessage(error) });
+    }
+  }
 });
 
-volunteerRouter.delete("/:volunteerID", async (req, res) => {
-  const { volunteerID } = req.params;
+/* Update volunteer status by:
+  - id, through URI (ex. /volunteers/1)
+  - userId, through query param (ex. /volunteers/?userId=1)
+*/
+volunteerRouter.put(
+  "/:volunteerId?",
+  volunteerDtoValidator,
+  async (req, res) => {
+    const { volunteerId } = req.params;
+    const { userId } = req.query;
+    const contentType = req.headers["content-type"];
+
+    if (volunteerId && userId) {
+      await sendResponseByMimeType(res, 400, contentType, [
+        {
+          error: "Cannot update by both volunteerId and userId",
+        },
+      ]);
+      return;
+    }
+    if (volunteerId) {
+      try {
+        await volunteerService.updateVolunteerById(volunteerId, {
```

Comment:
Why do we have both `updateVolunteerById` and `updateVolunteerByUserId`? When we are updating a volunteer, wouldnt we always have the id, then i think we prolly dont need `updateVolunteerByUserId`, but I am prolly missing some use case